### DISCUSSION TITLE
Add a keyword for modifying a value type (for example int) in synchronized manner

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/csharplang#5878] [Language Ideas]
Nowadays we have to use a mutex to do it, but it would be nice if I can just use a keyword then compiler helps me out.
```C#
checked locked(MyUInt32)
{
    if(MyDBHandle.Index == MyUInt32)
    {
        MyUInt32--;
    }
}
await MyThreadedRPC(MyDBHandle, MyUInt32);
```

**Comment by HaloFour** [comment]
Every reference type has a monitor, you can use that to synchronize access to anything, including reads/writes of value types using the existing `lock` keyword.

**Reply by Xyncgas** [reply]
But my UInt64 is a value type, why can't I just lock the integer it seems more obvious what I am trying to do, "I am trying to lock this integer so I am going to do something to it while others wait for me to finish"

**Reply by HaloFour** [reply]
Value types don't have monitors.  But that's irrelevant, you aren't "locking" the value you're trying to modify, you're locking an instance that represents the critical section.  What you modify within that critical section is entirely unrelated to what you are locking.

**Reply by CyrusNajmabadi** [reply]
What would you be locking on?

**Reply by bernd5** [reply]
If value types would have an implicit monitor we would need to increase all structs for that feature which is in 90 % just memory wasting.
In my eyes the same is true for object - most objects don't need such a monitor, but this was copied from java...

**Comment by TahirAhmadov** [comment]
Just create something like this:
```cs
int intValue = 123;
object locker = new();

lock(locker)
{
  if(intValue == 123) ++intValue;
}
```

**Reply by bernd5** [reply]
As long as not all threads have access to the same lock / object you would gain no safety by this locker...

**Reply by TahirAhmadov** [reply]
Then, you have to wrap the int value and the locker into a class and pass that around, to force all components to go through the locker.

**Reply by Xyncgas** [reply]
I know, I know, I just think maybe my little friend over here (compiler) can help me out, by creating a mutex / monitor for me to lock and release, when I've say so in the source code

**Reply by HaloFour** [reply]
> I know, I know, I just think maybe my little friend over here (compiler) can help me out, by creating a mutex / monitor for me to lock and release, when I've say so in the source code

The compiler does, when you use `lock` over a reference type.  You can't associate a mutex/monitor with a value type because it doesn't have an identity separate from its value.

**Reply by CyrusNajmabadi** [reply]
> I know, I know, I just think maybe my little friend over here (compiler) can help me out, by creating a mutex / monitor for me to lock and release, when I've say so in the source code

Waht would the mutex/monitor be assocaited with?  How do you envision the code-gen for this working?  

I'm concerned with this because it woudl be far too easy to accidentally lock over the wrong thing and hten not actually get the right behavior.

**Comment by bernd5** [comment]
Usually value types live on the stack so there is no sharing across multiple threads. This could be only the case with references or fields inside of a non value type...
In the later case the locking should be implemented in the field holder...

**Reply by Xyncgas** [reply]
You are correct about value types living on the stack so there is no point of racing, same with functional programming as there is no shared consequences

**Comment by CyrusNajmabadi** [comment]
Can you use Interlocked.Decrement?

**Reply by TahirAhmadov** [reply]
I think in the OP example, he has complicated logic - the integer is expected to stay the same through the if, the increment, and what not. But yes, in other cases, Interlocked.Increment/Decrement are the right solution.

**Reply by Xyncgas** [reply]
I am decreasing a pointer (UInt64) / Index to the last position about an array, if I am deleting the last element of this array, because I am making network calls I would await it which I let my users to invoke as often as they would like and expect their action has conveyed to results in order for a better user experience therefore these multiple invokes of my async method shall be syncrhonized in some way other wise I end up with the wrong index to the last position of my array as with all multi-threaded integer operations ```i--;``` or sometimes my RemoteRPC async method encounters an exception and didn't do anything but because of thread racing my index in the  database points to an element further than the actual counts of my documents therefore including in range and potentially leaking documents that was supposed to be deleted

This was just for fun, I typed this but I don't expect anyone to care about the details of my application
